### DISCUSSION TITLE
File path of Android with "content://" scheme

### DISCUSSION THREAD

**Discussion by tsctsc6** [AvaloniaUI/Avalonia#17080] [Q&A] [answered]
I have a app that use SQLite, the file path is:
```
public class MyDbService
{
    public static string DatabasePath => Path.Combine(AppContext.BaseDirectory, "my.db");
    //......
}
```
And I did the backup function:
```
private async Task BackupAsync()
{
    using var file = await App.Current.TopLevel!.StorageProvider
        .SaveFilePickerAsync(new FilePickerSaveOptions
    {
        Title = "Backup",
        ShowOverwritePrompt = true,
        DefaultExtension = "db",
        SuggestedFileName = "name",
        FileTypeChoices = [new FilePickerFileType("SQLite File")
        {
            Patterns = ["*.db"],
        }]
    });
    if (file is null) return;
    using var localDbFile = await App.Current.TopLevel!.StorageProvider
        .TryGetFileFromPathAsync(MyDbService.DatabasePath);
    if (localDbFile is null) return;
    await App.Current.DbService.CloseAsync();
    await using var localStream = await localDbFile.OpenReadAsync();
    await using var backupSteam = await file.OpenWriteAsync();
    await localStream.CopyToAsync(backupSteam);
    await backupSteam.FlushAsync();
}
```
I found that the app works well on Windows, but not on Android.

The point is `MyDbService.DatabasePath` , it is `/data/user/0/com.CompanyName.MyApp/files/my.db` on Android, and `StorageProvider.TryGetFileFromPathAsync(string)` return null.

Yes, now I understand that I can't use `StorageProvider.TryGetFileFromPathAsync(string)` but should use `StorageProvider.TryGetFileFromPathAsync(Uri)` on cross platform project. However, I don't know how to convert path string to path uri, especially Android seems to need "content://" scheme uri. The database api need path string and work well on Android. How to get "content://" scheme uri of `AppContext.BaseDirectory` ?

**Comment by maxkatz6** [comment]
Android implementation of `TryGetFileFromPathAsync(Uri)` doesn't specifically require "content://" scheme.
"file://" scheme should work well until you have read permission to the path.
The same also applies to `TryGetFileFromPathAsync(string)` method, which only converts input string into a "file://" URI and redirects call to `TryGetFileFromPathAsync(Uri)`.

**Reply by maxkatz6** [reply]
Also, that's important to understand, but your app _does not_ have full access to app installation folder, not by default at least.
You shouldn't use `Path.Combine(AppContext.BaseDirectory, "my.db")` as a config/database path there.

Consider using [Environment.GetFolderPath](https://learn.microsoft.com/en-us/dotnet/api/system.environment.getfolderpath?view=net-8.0) with `LocalApplicationData`.

**Comment by tsctsc6** [comment]
I try `Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)` and I found it is the same with `AppContext.BaseDirectory` on Android.

And I try `StorageProvider.TryGetFileFromPathAsync($"file://{PwdGenDbService.DatabasePath}")` and still returns null.

**Comment by tsctsc6** [comment, marked as answer]
Considering that the database API works fine, I guess I can use the lowest-level API: `FileStream` .
`await using var localStream = new FileStream(MyDbService.DatabasePath, FileMode.Open);`
It works!
